core/chains/terra/terratxm/txm.go

```diff
@@ -0,0 +1,355 @@
+		confirmMaxPolls:    100,
+		fallbackGasPrice:   sdk.NewDecCoinFromDec("uluna", fgp),
+		gasLimitMultiplier: gasLimitMultiplier,
+	}, nil
+}
+
+// Start subscribes to pg notifications about terra msg inserts and processes them.
+func (txm *Txm) Start() error {
+	return txm.starter.StartOnce("terratxm", func() error {
+		sub, err := txm.eb.Subscribe(pg.ChannelInsertOnTerraMsg, "")
+		if err != nil {
+			return err
+		}
+		txm.sub = sub
+		go txm.run(sub)
+		return nil
+	})
+}
+
+func (txm *Txm) confirmAnyUnconfirmed() {
+	// Confirm any broadcasted but not confirmed txes.
+	// This is an edge case if we crash after having broadcasted but before we confirm.
+	broadcasted, err := txm.orm.SelectMsgsWithState(Broadcasted)
+	if err != nil {
+		// Should never happen but if so, theoretically can retry with a reboot
+		txm.lggr.Errorw("unable to look for broadcasted but unconfirmed txes", "err", err)
+		return
+	}
+	if len(broadcasted) == 0 {
+		return
+	}
+	msgsByTxHash := make(map[string][]TerraMsg)
+	for _, msg := range broadcasted {
+		msgsByTxHash[*msg.TxHash] = append(msgsByTxHash[*msg.TxHash], msg)
+	}
+	for txHash, msgs := range msgsByTxHash {
+		err := txm.confirmTx(txHash, getIDs(msgs))
+		if err != nil {
+			txm.lggr.Errorw("unable to confirm broadcasted but unconfirmed txes", "err", err, "txhash", txHash)
+		}
+	}
+}
+
+func (txm *Txm) run(sub pg.Subscription) {
+	defer close(txm.done)
+	txm.confirmAnyUnconfirmed()
+	for {
+		select {
+		case <-sub.Events():
+			txm.sendMsgBatch()
+		case <-txm.ticker.C:
+			txm.sendMsgBatch()
+		case <-txm.stop:
+			return
+		}
+	}
+}
+
+func (txm *Txm) sendMsgBatch() {
+	unstarted, err := txm.orm.SelectMsgsWithState(Unstarted)
+	if err != nil {
+		txm.lggr.Errorw("unable to read unstarted msgs", "err", err)
+		return
+	}
+	if len(unstarted) == 0 {
+		return
+	}
+	if len(unstarted) > MaxMsgsPerBatch {
+		unstarted = unstarted[:MaxMsgsPerBatch+1]
+	}
+	txm.lggr.Debugw("building a batch", "batch", unstarted)
+	var msgsByFrom = make(map[string][]TerraMsg)
+	for _, m := range unstarted {
+		var ms wasmtypes.MsgExecuteContract
+		err := ms.Unmarshal(m.Msg)
+		if err != nil {
+			// Should be impossible given the check in Enqueue
+			txm.lggr.Errorw("failed to unmarshal msg, skipping", "err", err, "msg", m)
+			continue
+		}
+		m.ExecuteContract = &ms
+		_, err = sdk.AccAddressFromBech32(ms.Sender)
+		if err != nil {
+			// Should never happen, we parse sender on Enqueue
+			txm.lggr.Errorw("unable to parse sender", "err", err, "sender", ms.Sender)
+			continue
+		}
+		msgsByFrom[ms.Sender] = append(msgsByFrom[ms.Sender], m)
+	}
+
+	txm.lggr.Debugw("msgsByFrom", "msgsByFrom", msgsByFrom)
+	gp := txm.tc.GasPrice(txm.fallbackGasPrice)
+	for s, msgs := range msgsByFrom {
+		sender, _ := sdk.AccAddressFromBech32(s) // Already checked validity above
+		key, err := txm.ks.Get(sender.String())
+		if err != nil {
+			txm.lggr.Errorw("unable to find key for from address", "err", err, "from", sender.String())
```

Comment:
Hmm, this means you could halt the queue just because one key was missing. I wonder if we might actually want to continue rather than return here so we can at least send messages which DO have a present key?

Reply:
done